### ISSUE TITLE
Rerequest offer

### ISSUE THREAD

**Issue by kaladivo** [vexl-it/vexl#263]

**Comment by leomira** [comment]
duplicate of/connected to 
#311 
#292 
#299

**Comment by leomira** [comment]
also indirectly connected to #274

**Comment by kaladivo** [comment]
There are 3 actions should be implemented in the app:

1. Request offer
2. Cancel request
3. Request again

**when I request the offer**: 
- Other user should be notified
- I can not sent another request until interval has passed
- I can cancel the request - as long as other side has not accepted
- Text with how long until I can request again is visible on offer screen

**When I cancel the offer**
- Other user should not be notified via push notification
- Other user could not accept the offer again
- I can rerequest the offer in the next day again

**When I rerequest the offer**
- Other user is notified
- Another message is send and displayed after accepting
- Rest is the same as request

- Offer is returning to colored state (in marketplace list) after request timeout period and that is even in the case that the offer was not canceled or accepted.

Questions: 
- What should the rerequest interval be - for now every day should user request offer once. Can be subject to change in the future. 

Create QA test case.

**Comment by kaladivo** [comment]
Changes proposed: 
**BE:** 
- Chat service is modified so requests can be sent again after some time period
  - New field `request_date` is added to request in db (day resolution, not time to preserve privacy)
- New `cancelled` message type is added to message type - TODO - here might be compatibility problem with the old app.

**FE**
- Changes in offer detail UI
- Changes in chat detail in request pending state UI
- Changes in logic of making offers black/white in marketplace list / offer detail

**Comment by kaladivo** [comment]
BE changeset: https://github.com/vexl-it/vexl-backend/commit/b87e300a5d7881af08eebf57357bdae0892346e4

**Comment by kaladivo** [comment]
Chat can have following states: 
```
enum([
  'initial',
  'requested',
  'denied',
  'cancelled',
  'accepted',
  'finished' // already interacted but exited chat
])
```
**questions**?
- In which states should the `days until rerequest` text should be displayed? 
- If other side cancels the request, how should it be displayed to offer's autor?
  - proposal: request screen with action button `remove request` in place where accept/deny buttons were
- Shall we remove user from whitelist once he leaves the chat? What should happen to other side chat view in this case?

**Comment by kaladivo** [comment]
Specification was refined since the scope of this change is larger than was expected at first.

**User story:**
- User should be able to rerequest offer every day (can be subject to change, configurable in backend and through firebase remote config)
- User should be able to cancel pending request that he/she made
- When the one who is not creator leaves the offer and requests the offer again: 
  -  Other side should see request in chat timeline action bar should be used for this feature (similar to identity reveal)
- When the one who is creator leaves the chat
  - other side should be able to rereques offer
- When rerequesting offer that user previously interacted with but leaved the chat, it should be communicated with the user, that some interaction with this offer was made

**Technical changes:**

_EAT units used. One ETA unit is halfday. ETA units are rounded up to be integers._


- [x] Allow user to request offer every X days on backend - finished in  https://github.com/vexl-it/vexl-backend/commit/b87e300a5d7881af08eebf57357bdae0892346e4
- [x] When user leaves the chat, backend should remove all records about two sides ever being in contact. The record from whitelist table should be removed. - ETA 1 - done in https://github.com/vexl-it/vexl-backend/commit/5a33728aa828406969f454a2a65a48cb31e410be
- [x] Implement UI for - user is able to rerequest the offer every X day - take the value from firebase remote config - mostly finished #354 
- [x] Implement UI for - User has received request but later that request was cancelled - Instead of accept/decline user should be informed about cancellation of the request and be able to erase the chat - ETA 2
- [ ] Implement UI for - User who responded to offer has left the chat and later rerequested the offer again - ETA 2
  - Creator sees the request in similar way as reveal identity (action bar)
  - The one who is responding to offer knows that this is not his/her first interaction with the offer
- [ ] Implement UI for - User who has created the offer has left the chat - ETA 2
  - User who has responded to the offer should be able to rerequest the offer from chat and from offers list /offer detail 
  - User who has created offer should see that this is not the first interaction with other side **(??)**
- [ ] Make sure UI make sense for - both users delete chat and offer is rerequested - ETA 2
  - Both sides should see, that this is not their first interaction - while sending request / while accepting request
- [ ] Write comprehensive test cases for all scenarios, so we know how to test this feature properly. - ETA 1

**Comment by leomira** [comment]
We have design where we indicate if it is first or repeating interaction with the offer - it is the offer detail/request. Why not to keep this?

**Comment by kaladivo** [comment]
Flow when author of the offer receives another rerequest, shoud look like this
https://github.com/vexl-it/vexl/assets/11568934/2425376c-8bf5-4194-802d-075960082ce3

**Comment by kaladivo** [comment]
Additional todo: 
- [x] implement new type of notification that does not deliver any system notification but makes the app refresh specific inbox - this should be used when other side cancels messaging request - finished here: https://github.com/vexl-it/vexl-backend/pull/4/commits/89cc3327b409e0ad1779fd6fe6e630c1545bade9

**Comment by kaladivo** [comment]
We'll have to find a better way how to communicate with user on the marketplace list screen that the offer was either left by one side or request was cancelled. 

See following screenshot: 
![Snímek obrazovky 2023-07-07 v 14 41 06](https://github.com/vexl-it/vexl/assets/11568934/fe7dbf40-211b-4400-9daf-04c076ac4f07)


There are multiple states in which user can rerequest the offer again: 
1. Chat was leaved by one side
2. Request offer was cancelled
3. Request was declined

In each of these states, the offer will stay at the bottom in "interacted state". User will be informed about these states on request detail, but there is not enough space (in the current design at least) to communicate all cases on marketplace screen.

**Comment by kaladivo** [comment]
# Rerequest test case


**Before every test case** 
Before every test case the following state should be set. 

- Login user A and User B
- User A and B import the same contact
- User A should create an offer
- User B should request the offer


## Test cases
_On stage, there is no limit on sending requests. User should be able to rerequest immediately._
### Rerequest

- User A should see message about rerequest possible in one day
- After one day user B should be able to rerequest the offer
- User A should get a notification about rerequest – TODO
- User A should be able to accept the offer

### Rerequest after rejecting

- User A denies request from user B
- Wait one day
- User B requests the offer again
- User A accepts and the chat is open

### Rerequest after one side leaves

- User A accepts the request
- User A deletes/leaves the chat
- User B sees: 
- User A has left
- User A sees:
  - On offer detail it is communicated that this is not the first interaction
  - Offer is at the end of the list in marketplace and marked as closed
- User A rerequests the offer
- User B sees
  - An indication that there were some interaction with that user before
  - Button to access previous conversation
- User B accepts
- Both sides can message again

### No rerequest after blocking

- User A blocks other user
- User B does see the offer as pending
- When the user B tries to rerequest the offer an error message with 404 is displayed (offer or does not exist, or you are blocked by the user)

**Comment by kaladivo** [comment]
Fix copy: 
rerequest -> request again
leaved -> left

Put state of the request on marketplace item view.

**Comment by kaladivo** [comment]
Few more refinements: 
- There are a few UI states when user rerequests the offer (marketplace offer) - **Ordering of the offer is not changed when user interacts with the offer**
  - user has not interacted with offer yet - white background, _request_ button
  - user has interacted with the offer and request again can NOT be sent - black background,  _go to chat_ button
  - User hash interacted with the offer and request can be sent - black background  _request again_ button
- Rerequest can be sent from chat detail
- When author of the offer leaves the chat the other side see button to request the offer again.
